### PR TITLE
[#623] hard-limit large connection timeouts

### PR DESCRIPTION
**Pull request by d-w-moore**
Prevent PRC from blowing up when excessive socket timeout values are used.

This is  more of a Python or libc underlying problem from the  but we want to be proactive and cap the value to its maximum practicable value for the benefit of projects relying on PRC.

### CODE REVIEW

File: irods/__init__.py
Comment (trel) [inline irods/__init__.py:45]: I'm not convinced we have to keep it so close to the actual MAX of a 64bit signed int of nanoseconds.

We can set it to a much smaller (read: resonable) max.  Perhaps only 100 years(?!).

Comment (d-w-moore) [reply irods/__init__.py:45]: We could also just use the integer part, which would reduce the verbosity of the test... is there a concern about using a maximum limit too close to the system's actual limit?

Comment (trel) [reply irods/__init__.py:45]: Agreed - integer part seems fine

File: irods/session.py
Comment (trel) [inline irods/session.py:28]: let's switch the order here, alpha order AND the order they appear in `__init__.py`

File: irods/test/connection_test.py
Comment (trel) [inline irods/test/connection_test.py:174]: what do these do?

Comment (d-w-moore) [reply irods/test/connection_test.py:174]: This is just there to use connection and prove it's still  viable.  It is probably irrelevant now that the timeout takes immediate effect in the value of the connection timeout attribute, rather than showing up lazily in the act of using the connection.  Will delete....

File: irods/session.py
Comment (alanking) [inline irods/session.py:145]: Should we continue to set `_cached_connection_timeout` here? It looks like the `configure()` method is setting the connection pool timeout to `None` if `_cached_connection_timeout` is not set, and we call `configure` just a few lines below here in the default case.

If we should NOT be setting `_cached_connection_timeout` here, perhaps the reason should go in a comment here? Or, better yet, is there a reason to keep `_cached_connection_timeout` at all?

File: irods/test/connection_test.py
Comment (alanking) [inline irods/test/connection_test.py:171]: Some minor paranoia here...

Perhaps we should set the timeout to some good value in between these two assertions in order to see that the setter took effect. Or, we could split the test up into one for the `__init__` function (i.e. `make_session`) and one for the setter.

The reason I am suggesting this is that we don't know that the `connection_timeout` is `MAXIMUM_CONNECTION_TIMEOUT` here because the setter used a too-large value. It could just be unchanged from when the `make_session` call used a too-large value. Not that there is a difference in this case...

Hopefully that makes sense.

**Review by alanking — COMMENTED** [review]
Had a couple of questions/concerns but the approach seems good

File: irods/session.py
Comment (d-w-moore) [reply irods/session.py:145]: The session attribute `_cached_connection_timeout` exists as a place to store the current session timeout value, whether or not a resident pool has been created. It's possible there is some redundancy here, I'll take a look at it.

File: irods/test/connection_test.py
Comment (d-w-moore) [reply irods/test/connection_test.py]: Yes, I see, I think... asserting the sess.connection_timeout value is really asserting the pool's value.  That is certainly not what is intended.

File: irods/session.py
Comment (d-w-moore) [reply irods/session.py]: Went ahead with initializing `_cached_connection_timeout` in the session c'tor, because it makes sense to.

File: irods/test/connection_test.py
Comment (alanking) [reply irods/test/connection_test.py]: Well... I actually was thinking about the fact that we are using the setter to set the value of `connection_timeout` and asserting that it then has the value that it already holds. The reason I have a problem with this is because the setter may just be leaving the value the way it was.

From this test, we cannot tell what it did. Did it set `connection_timeout` to some maximum value (as we are doing), or did it leave the value unchanged? That's why I'm suggesting setting it to some other, valid value in between so that we can see that the value actually was changed by the setter.

Something like this was what I had in mind:
```python
        # This sets the original value using the __init__ method - should be the "max" value.
        sess = helpers.make_session(connection_timeout = MAXIMUM_CONNECTION_TIMEOUT + 1)
        self.assertEqual(sess.connection_timeout, MAXIMUM_CONNECTION_TIMEOUT)
        # Then we set the value to something different...
        sess.connection_timeout = 1 # (or some other valid value)
        self.assertEqual(sess.connection_timeout, 1)
        # Now we set the invalid value, asserting that the setter is using the "max" value
        # and this ensures that it's not just unchanged from its previous value.
        sess.connection_timeout = MAXIMUM_CONNECTION_TIMEOUT + 1
        self.assertEqual(sess.connection_timeout, MAXIMUM_CONNECTION_TIMEOUT)
```

Comment (d-w-moore) [reply irods/test/connection_test.py]: Ok think I've done it.... @alanking , please examine

File: irods/test/connection_test.py
Comment (alanking) [inline irods/test/connection_test.py:175]: Would it be more appropriate to check what the value *is* equal to (would it be 1?) rather than what it is *not* equal to?

File: irods/test/connection_test.py
Comment (alanking) [reply irods/test/connection_test.py]: Yes, I think so. I've left a comment on the change, so I'll resolve this one as the major concern has been addressed. Thanks!

File: irods/session.py
Comment (alanking) [reply irods/session.py]: Please resolve if this has been addressed. It seems to me that it has. Thanks

File: irods/test/connection_test.py
Comment (d-w-moore) [reply irods/test/connection_test.py:175]: But testing equality will be just as good.

**Review by alanking — APPROVED** [review]
\# it

**Review by korydraughn — APPROVED** [review]
Let's wait for one more approval before pounding.

**Comment by korydraughn** [comment]
Ignore my last comment.

Yep, pound it!

**Pull request merged**
